### PR TITLE
Refactored Drafty

### PR DESCRIPTION
**Pull request by or-else**
Drafty refactored to get rid of icons and a few other changes.
Added tests.

### CODE REVIEW

**Comment by or-else** [comment]
The merge conflicts in umd/tinode.dev.js and umd/tinode.prod.js are expected, not an issue.
The merge conflict in src/drafty.js is weird as I cannot make it go away in github. I'll resolve it manually when I merge.

File: src/drafty.js [hunk 1]
```diff
@@ -1698,242 +1459,466 @@ Drafty.attrValue = function(style, data) {
...
+    chunks.push({
+      txt: line.slice(start, end)
+    });
+  }
+
+  return chunks;
+}
+
+// Inverse of chunkify: iterates spans bottom-up. Returns a tree of formatted spans.
+function iterateSpans(line, start, end, spans, formatter, context) {
```
Comment (aforge) [inline src/drafty.js:1517]: Rename `formatter` -> `defaultFormatter`?

File: src/drafty.js [hunk 2]
```diff
@@ -1698,242 +1459,466 @@ Drafty.attrValue = function(style, data) {
...
+
+  // Sort spans first by start index (asc) then by length (desc).
+  spans.sort((a, b) => {
+    const diff = a.at - b.at;
+    return diff != 0 ? diff : b.len - a.len;
+  });
+
+  // Drop invalid spans and staggered spans: the second and subsequent
+  // spans when they overlap like '_first *second_ third*'.
+  let end = -2;
```
Comment (aforge) [inline src/drafty.js:1728]: Any specific reason for -2? Attachments?
If so please consider Number.NEGATIVE_INFINITY to make it more readable.

File: src/drafty.js [hunk 3]
```diff
@@ -1698,242 +1459,466 @@ Drafty.attrValue = function(style, data) {
...
-  };
-
-  if (typeof txt == 'string') {
-    preview.txt = txt.substr(0, length);
+  if (sp) {
+    if (sp.tp == 'QQ') {
+      // Skip quoted text
+      return null;
+    }
+    if (sp.tp == 'BR' && at == 0) {
```
[the remark this answers is not in the document]
Comment (or-else) [reply src/drafty.js:1864]: Line breaks are converted to spaces. There is no value in a leading space.

File: src/drafty.js [hunk 4]
```diff
@@ -438,372 +443,152 @@ const DECORATORS = {
  * @class Drafty
  * @constructor
  */
-const Drafty = function() {}
+const Drafty = function() {
+  this.txt = '';
+  this.fmt = [];
```
[the remark this answers is not in the document]
Comment (or-else) [reply src/drafty.js:448]: They are removed here when serialization happens:
https://github.com/tinode/tinode-js/blob/be043c46971a4fec9810934ec24bd772876d7099/src/tinode.js#L294

File: src/drafty.js [hunk 5]
```diff
@@ -1698,242 +1459,466 @@ Drafty.attrValue = function(style, data) {
...
-        }
-      }
+
+  if (children) {
+    children.forEach((c) => {
+      previewFormatter.call(this, c.sp, c.txt, c.children);
     });
+  } else if (txt) {
+    if (at + txt.length > this.maxLength) {
+      this.drafty.txt += txt.slice(0, this.maxLength - at - 1) + '…';
```
[the remark this answers is not in the document]
Comment (or-else) [reply src/drafty.js:1876]: 1. This code works correctly for the RTL languages: https://jsfiddle.net/682peayb/
2. Tinode is broken for RTL anyway. We need dir="auto" for all fields where RTL text may appear.

File: src/drafty.js (hunk 1 above)
Comment (or-else) [reply src/drafty.js:1517]: done

File: src/drafty.js (hunk 2 above)
Comment (or-else) [reply src/drafty.js:1728]: Used Number.MIN_SAFE_INTEGER

**Comment by or-else** [comment]
All seems to work.

File: src/drafty.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aforge) [reply src/drafty.js:1876]: screen

The current handling of images in quoted sections still treats `<BR>` as line breaks. Is it expected?

**Review by aforge — APPROVED** [review]

File: src/drafty.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (or-else) [reply src/drafty.js]: done

**Pull request merged**